File: HoTT-Model/src/main/java/gde/model/voice/VoiceFile.java
package gde.model.voice;

import gde.model.enums.TransmitterType;
import javafx.beans.InvalidationListener;
import javafx.beans.binding.BooleanBinding;
import javafx.beans.binding.IntegerBinding;
import javafx.beans.property.IntegerProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.FXCollections;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;

/**
 * Representation of a void data file (.vdf).
 *
 * @author oliver.treichel@gmx.de
 */
public class VoiceFile {
    private final class DirtyBinding extends BooleanBinding {
        @Override
        protected boolean computeValue() {
            final VoiceFile voiceFile = VoiceFile.this;
            return !voiceFile.equals(voiceFile.savedState);
        }
    }

    private final class SizeBinding extends IntegerBinding {
        @Override
        protected int computeValue() {
            return voiceData.size();
        }
    }

    private final VoiceFile savedState;
    private final ObjectProperty<CountryCode> country = new SimpleObjectProperty<>(CountryCode.GLOBAL);
    private final ObjectProperty<TransmitterType> transmitterType = new SimpleObjectProperty<>();
    private final ObjectProperty<VDFType> vdfType = new SimpleObjectProperty<>();
    private final IntegerProperty vdfVersion = new SimpleIntegerProperty();
    private final ObservableList<VoiceData> voiceData;
    private final IntegerBinding sizeBinding;
    private final BooleanBinding dirtyProperty;

    private VoiceFile() {
        voiceData = FXCollections.observableArrayList();
        sizeBinding = null;
        dirtyProperty = null;
        savedState = null;
    }

    public VoiceFile(final VDFType vdfType, final TransmitterType transmitterType, final int vdfVersion, final CountryCode country) {
        this(vdfType, transmitterType, vdfVersion, country, FXCollections.observableArrayList());
    }

    public VoiceFile(final VDFType vdfType, final TransmitterType transmitterType, final int vdfVersion, final CountryCode country,
            final ObservableList<VoiceData> voiceData) {
        sizeBinding = new SizeBinding();
        dirtyProperty = new DirtyBinding();

        setVdfType(vdfType);
        setTransmitterType(transmitterType);
        setVdfVersion(vdfVersion);
        setCountry(country);

        this.voiceData = voiceData;
        this.voiceData.addListener((ListChangeListener<VoiceData>) c -> sizeBinding.invalidate());
        savedState = new VoiceFile();
        clean();

        this.country.addListener((a, b, c) -> dirtyProperty.invalidate());
        this.transmitterType.addListener((a, b, c) -> dirtyProperty.invalidate());
        this.vdfType.addListener((a, b, c) -> dirtyProperty.invalidate());
        this.vdfVersion.addListener((a, b, c) -> dirtyProperty.invalidate());
        this.voiceData.addListener((ListChangeListener<VoiceData>) c -> dirtyProperty.invalidate());
        this.voiceData.addListener((InvalidationListener) c -> dirtyProperty.invalidate());
    }

    public VoiceFile(final VDFType vdfType, final TransmitterType transmitterType, final int vdfVersion, final int countryCode) {
        this(vdfType, transmitterType, vdfVersion, countryCode, FXCollections.observableArrayList());
    }

    public VoiceFile(final VDFType vdfType, final TransmitterType transmitterType, final int vdfVersion, final int countryCode,
            final ObservableList<VoiceData> voiceData) {
        this(vdfType, transmitterType, vdfVersion, CountryCode.forCode(countryCode), voiceData);
    }

    public void clean() {
        if (savedState != null) savedState.copy(this);
        if (dirtyProperty != null) dirtyProperty.invalidate();
    }

    public void copy(final VoiceFile other) {
        if (other == this) return;

        setCountry(other.getCountry());
        setTransmitterType(other.getTransmitterType());
        setVdfType(other.getVdfType());
        setVdfVersion(other.getVdfVersion());
        voiceData.clear();
        other.getVoiceData().forEach(v -> voiceData.add(new VoiceData(v.getName(), v.getRawData())));
        clean();
    }

    public ObjectProperty<CountryCode> countryProperty() {
        return country;
    }

    public BooleanBinding dirtyProperty() {
        return dirtyProperty;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass()) return false;
        final VoiceFile other = (VoiceFile) obj;
        if (country.get() == null) {
            if (other.country.get() != null) return false;
        } else if (!country.get().equals(other.country.get())) return false;
        if (transmitterType.get() == null) {
            if (other.transmitterType.get() != null) return false;
        } else if (!transmitterType.get().equals(other.transmitterType.get())) return false;
        if (vdfType.get() == null) {
            if (other.vdfType.get() != null) return false;
        } else if (!vdfType.get().equals(other.vdfType.get())) return false;
        if (vdfVersion.get() != other.vdfVersion.get()) return false;
        if (voiceData == null) {
            if (other.voiceData != null) return false;
        } else if (!voiceData.equals(other.voiceData)) return false;
        return true;
    }

    public CountryCode getCountry() {
        return country.get();
    }

    public int getDataSize() {
        return voiceData.stream().mapToInt(vd -> vd.getRawData().length).sum();
    }

    public TransmitterType getTransmitterType() {
        return transmitterType.get();
    }

    public VDFType getVdfType() {
        return vdfType.get();
    }

    public int getVdfVersion() {
        return vdfVersion.get();
    }

    public int getVoiceCount() {
        return voiceData.size();
    }

    public ObservableList<VoiceData> getVoiceData() {
        return voiceData;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + (country.get() == null ? 0 : country.get().hashCode());
        result = prime * result + (transmitterType.get() == null ? 0 : transmitterType.get().hashCode());
        result = prime * result + (vdfType.get() == null ? 0 : vdfType.get().hashCode());
        result = prime * result + vdfVersion.get();
        result = prime * result + (voiceData == null ? 0 : voiceData.hashCode());
        return result;
    }

    public void setCountry(final CountryCode country) {
        this.country.set(country);
    }

    public void setTransmitterType(final TransmitterType transmitterType) {
        this.transmitterType.set(transmitterType);
    }

    public void setVdfType(final VDFType vdfType) {
        this.vdfType.set(vdfType);
    }

    public void setVdfVersion(final int vdfVersion) {
        this.vdfVersion.set(vdfVersion);
    }

    public ObjectProperty<TransmitterType> transmitterTypeProperty() {
        return transmitterType;
    }

    public ObjectProperty<VDFType> vdfTypeProperty() {
        return vdfType;
    }

    public IntegerProperty vdfVersionProperty() {
        return vdfVersion;
    }

    public IntegerBinding voiceCountProperty() {
        return sizeBinding;
    }
}

File: HoTT-Voice/src/main/kotlin/de/treichels/hott/voice/VoiceFile.kt
package de.treichels.hott.voice

import de.treichels.hott.model.enums.TransmitterType
import javafx.beans.InvalidationListener
import javafx.beans.Observable
import javafx.beans.binding.BooleanBinding
import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.collections.FXCollections
import tornadofx.*

data class VoiceInfo(val vdfType: VDFType, val flashSize: Int, val sectorCount: Int, val sectorSize: Int, val voiceVersion: Int, val infoSize: Int, val maxDataSize: Int)
data class VoiceFileInfo(val vdfType: VDFType, val size: Int, val baseAddress: Int, val index: Int, val reverseIndex: Int, val sampleRate: Int, val name: String)

open class ObservableBase : Observable {
    // listener list
    private val listeners = mutableListOf<InvalidationListener>()

    override fun addListener(listener: InvalidationListener?) {
        if (listener != null) listeners.add(listener)
    }

    override fun removeListener(listener: InvalidationListener?) {
        if (listener != null) listeners.remove(listener)
    }

    fun invalidate() {
        listeners.forEach { it.invalidated(this) }
    }
}

open class VoiceFileBase(vdfType: VDFType, transmitterType: TransmitterType, vdfVersion: Int, country: CountryCode, list: List<VoiceData>) : ObservableBase(){
    // properties
    val vdfTypeProperty = SimpleObjectProperty<VDFType>(vdfType)
    val transmitterTypeProperty = SimpleObjectProperty<TransmitterType>(transmitterType)
    val vdfVersionProperty = SimpleIntegerProperty(vdfVersion)
    val countryProperty = SimpleObjectProperty<CountryCode>(country)
    val voiceList = list.asObservable()
    val sizeProperty = voiceList.sizeProperty
    val systemVDFProperty: BooleanBinding = vdfTypeProperty.isEqualTo(VDFType.System)

    // simple values
    var vdfType: VDFType by vdfTypeProperty
    var transmitterType: TransmitterType by transmitterTypeProperty
    var vdfVersion by vdfVersionProperty
    var country: CountryCode by countryProperty
    val rawDataSize: Int
        get() = voiceList.stream().mapToInt { it.rawData.size }.sum()
    val size by sizeProperty
    val isSystemVDF by systemVDFProperty

    // copy constructor
    constructor(other: VoiceFileBase) : this(other.vdfType, other.transmitterType, other.vdfVersion, other.country, ArrayList(other.voiceList.map(VoiceData::clone)))

    init {
        // add invalidation listeners - trigger events on bindings
        this.countryProperty.addListener { _ -> invalidate() }
        this.transmitterTypeProperty.addListener { _ -> invalidate() }
        this.vdfTypeProperty.addListener { _ -> invalidate() }
        this.vdfVersionProperty.addListener { _ -> invalidate() }
        this.voiceList.addListener { _: Observable -> invalidate() }
    }

    // copy state
    fun copy(other: VoiceFileBase) {
        this.vdfType = other.vdfType
        this.transmitterType = other.transmitterType
        this.vdfVersion = other.vdfVersion
        this.country = other.country
        voiceList.clear()
        voiceList.addAll(other.voiceList.map(VoiceData::clone))
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is VoiceFileBase) return false

        if (vdfType != other.vdfType) return false
        if (transmitterType != other.transmitterType) return false
        if (vdfVersion != other.vdfVersion) return false
        if (country != other.country) return false
        if (voiceList != other.voiceList) return false

        return true
    }

    override fun hashCode(): Int {
        var result = vdfType.hashCode()
        result = 31 * result + transmitterType.hashCode()
        result = 31 * result + vdfVersion
        result = 31 * result + country.hashCode()
        result = 31 * result + voiceList.hashCode()
        return result
    }
}

/**
 * Representation of a void data file (.vdf).
 *
 * @author oliver.treichel@gmx.de
 */
class VoiceFile(vdfType: VDFType = VDFType.User, transmitterType: TransmitterType = TransmitterType.mc28, vdfVersion: Int = 3000, country: CountryCode = CountryCode.GLOBAL,
                list: MutableList<VoiceData> = mutableListOf()) : VoiceFileBase(vdfType, transmitterType, vdfVersion, country, FXCollections.observableList(list)) {

    private val savedState: VoiceFileBase = VoiceFileBase(this)
    val dirtyBinding: BooleanBinding = booleanBinding(this, this, op = { !equals(savedState) })
    val isDirty by dirtyBinding

    fun clean() {
        savedState.copy(this)
        dirtyBinding.invalidate()
    }
}
